agentcfg/fetch.go

```diff
@@ -121,3 +182,49 @@ func containsAnyPrefix(s string, prefixes []string) bool {
 	}
 	return false
 }
+
+type DirectFetcher struct {
+	cfgs []config.AgentConfig
+}
+
+func NewDirectFetcher(cfgs []config.AgentConfig) *DirectFetcher {
+	return &DirectFetcher{cfgs}
+}
+
+// Fetch finds a matching AgentConfig based on the received Query. Order of precedence:
+// - service.name and service.environment match a AgentConfig
+// - service.name matches a AgentConfig, service.environment == ""
+// - service.environment matches a AgentConfig, service.name == ""
+// Return an empty result if no matching result is found.
+func (f *DirectFetcher) Fetch(_ context.Context, query Query) (Result, error) {
+	name, env := query.Service.Name, query.Service.Environment
+	result := zeroResult()
+	var nameConf, envConf *config.AgentConfig
+
+	for i, cfg := range f.cfgs {
+		if cfg.Service.Name == name && cfg.Service.Environment == env {
+			nameConf = &f.cfgs[i]
+			break
+		} else if cfg.Service.Name == name && cfg.Service.Environment == "" {
+			nameConf = &f.cfgs[i]
+		} else if cfg.Service.Name == "" && cfg.Service.Environment == env {
+			envConf = &f.cfgs[i]
+		}
```

Comment:
As mentioned elsewhere, we'll also need a catchall case: where `cfg.Service.Name == "" && cfg.Service.Environment == ""`. That will have lowest precedence.

Reply:
Wouldn't that result in an empty config? In which case, the initial assignment `result := zeroResult()` would cover it?

Reply:
Read your comment [above](https://github.com/elastic/apm-server/pull/5177#discussion_r627238743), will wait for clarification